factor/totp/classes/factor.php

```diff
@@ -175,11 +175,25 @@ public function login_form_validation($data) {
             $totp = TOTP::create($factor->secret);
             if ($totp->verify($data['verificationcode'], time(), 1)) {
                 $result = array();
+                $this->update_lastlogon($factor);
             }
         }
         return $result;
     }
 
+    /**
+     * When validation code is correct - update last logon field for given factor.
+     *
+     * @param factor
+     * @return void
+     * @throws \dml_exception
+     */
+    public function update_lastlogon($factor) {
```

Comment:
I think this will end up being common to all factors, can you please move this back to the base class

Reply:
I'm also thinking it would be best to rename it. Its a subtle distinction but technically we are not logging on, we are verifying a factor. You could require 3 factors and pass 2 but fail 1 and so not actually login.

Likewise when you setup a new factor you are also verifying it but not logging in. So can you:

1) move to base class. OR at least put it in the interface and leave it here
2) rename to update_lastverified
3) because we have a common api in get_all_user_factors() I think it's best to move the {factor_totp} table into {tool_mfa}. I'm in two minds about this, either the api isn't right or the table is in the wrong spot but I'm not sure which is better. Even things like iprange will still update the verify date so I think it needs to be shared.